### ISSUE TITLE
Not all references to onDisconnect have been removed

### ISSUE THREAD

**Issue by boogerlad** [balderdashy/sails#2523]
<pre><code>warn: Socket disconnected, but session could not be loaded to pass to configured disconnect handler: `sails.config.sockets.onDisconnect()`.  Will pass a fake, empty session as argument to lifecycle callback.  Details:
 Error: Session could not be loaded
    at _createError (/root/cutequickposweb/node_modules/sails/lib/hooks/session/index.js:271:21)
    at Object._onImmediate (/root/cutequickposweb/node_modules/sails/lib/hooks/session/index.js:274:13)
    at processImmediate [as _immediateCallback] (timers.js:354:15) { [Error: Session could not be loaded] code: 'E_SESSION' }
</code></pre>

Or maybe I'm wrong and there's something wrong with my code? I followed the migration guide.

**Comment by sgress454** [comment]
Interesting... @mikermcneil?

**Comment by boogerlad** [comment]
I've updated to the latest version again. New messages:

<pre><code>warn: Socket disconnected, but session could not be loaded to pass to configured disconnect handler: `sails.config.sockets.onDisconnect()`.  Will pass a fake, empty session as argument to lifecycle callback.  Details:
 Error: Session could not be loaded
    at _createError (/root/cutequickposweb/node_modules/sails/lib/hooks/session/index.js:271:21)
    at Object._onImmediate (/root/cutequickposweb/node_modules/sails/lib/hooks/session/index.js:274:13)
    at processImmediate [as _immediateCallback] (timers.js:354:15) { [Error: Session could not be loaded] code: 'E_SESSION' }
warn: Socket disconnected, but session could not be loaded to pass to configured disconnect handler: `sails.config.sockets.onDisconnect()`.  Will pass a fake, empty session as argument to lifecycle callback.  Details:
 Error: Could not parse session id from cookie of connecting socket, and then failed again when trying to use a generated cookie. Something has probably gone wrong with your session store configuration.
    at _createError (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/lib/load-session-from-socket.js:43:17)
    at loadSessionFromSocket (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/lib/load-session-from-socket.js:46:7)
    at Socket.onSocketDisconnect (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/lib/on-connect.js:43:7)
    at Socket.emit (events.js:95:17)
    at Socket.emit (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/node_modules/socket.io/lib/socket.js:129:10)
    at Socket.onclose (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/node_modules/socket.io/lib/socket.js:410:8)
    at Client.onclose (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/node_modules/socket.io/lib/client.js:230:12)
    at Socket.emit (events.js:117:20)
    at Socket.onClose (/root/cutequickposweb/node_modules/sails/node_modules/sails-hook-sockets/node_modules/socket.io/node_modules/engine.io/lib/socket.js:248:10)
    at WebSocket.g (events.js:180:16) { [Error: Could not parse session id from cookie of connecting socket, and then failed again when trying to use a generated cookie. Something has probably gone wrong with your session store configuration.] code: 'E_SESSION' }
</code></pre>

**Comment by boogerlad** [comment]
I seem to have fixed it by removing the `afterDisconnect` function completely. I wasn't using it anyways. The stub in the migration guide is problematic then.

**Comment by prisonier** [comment]
I also have the same problem with same solution

**Comment by cyrilchapon** [comment]
Hello,

+1 !
Just pushing this one up, since I personnaly do use the callback and am not realy up to remove it =/

Thanks =)

**Comment by yossavi** [comment]
+1

**Comment by tjwebb** [comment]
@boogerlad @cyrilchapon would one of you mind submitting a PR with a fix for this?

**Comment by frwickst** [comment]
Please open this issue again.

I use the function and would really like this to be fixed. Removing a function that should work is not a proper fix.

**Comment by mateeyow** [comment]
:+1:

**Comment by kennysmoothx** [comment]
+1

**Comment by CiscoKidxx** [comment]
+1

Best,
Chris Newell

**Comment by richdunajewski** [comment]
:+1:

@sgress454 @mikermcneil 
Seems like this occurs when the session in Mongo (at least for me) expires and Mongo cleans up the record. Then, if the socket disconnects, Sails tries to dig up the session information from Mongo which has just been deleted, and voila we can't load the session.

I have my session cookies set to expire in 6 minutes, so when the client-side code detects that we're no longer logged in, it redirects the user to the login page. When the socket disconnects to perform the redirect, Sails attempts to pass the session info to this disconnect handler and you can see we have a problem. To the user, it makes absolutely no difference as they're navigating away. However the server is throwing a hissy when it comes up empty for the session data.

So then what should happen if your session expires in the DB? I think it's fine to pass an empty session object down to the disconnect handler but that error message needs to be suppressed. Even though it logs with a `warn` level, the stack trace is adding ~18 lines of nonsense to my logs each time a socket with expired session disappears. 

Maybe set this to a lower severity so it doesn't pollute the logs? Is there a more graceful way to handle this?

**Comment by mbdwey** [comment]
+1

**Comment by wedneyyuri** [comment]
+1. How can I use `afterDisconnect` handler?

**Comment by billyshena** [comment]
Hello guys, has there been any fixes to the afterDisconnect function?

**Comment by Guiiks** [comment]
+1 for a fix!

**Comment by tomsun** [comment]
Reproducible here too

**Comment by sgress454** [comment]
Hey all, it's not clear to me whether there's an actual issue here or if folks are just concerned about seeing the warning in their log.  What @richdunajewski said is spot on; if a session is cleared before the socket that was using that session disconnects, you're gonna see this warning.  If you want to keep the session alive, you can implement some sort of heartbeat in your front-end code that pings the server on an interval.  

> I use the function and would really like this to be fixed. Removing a function that should work is not a proper fix.

Please [read the migration guide](http://sailsjs.org/version-notes/0point10-to-0point11-migration-guide) and use `afterDisconnect` instead of `onDisconnect.

> How can I use afterDisconnect handler?

Replace `onDisconnect` with `afterDisconnect` and add a callback argument.  The differences are laid out in [the migration guide](http://sailsjs.org/version-notes/0point10-to-0point11-migration-guide).

> Hello guys, has there been any fixes to the afterDisconnect function?

As far as I can tell there aren't any issues with the afterDisconnect function.  If you're seeing an actual problem in your app that you think is directly related to disconnecting a socket, please post here.  If it's just about the warning, I'll go ahead and close this.

**Comment by billyshena** [comment]
@sgress454 : Thanks for your answer ! However, is there a way to shut down the warning? Because it's flooding the entire log stack of our app in production.

Best,

**Comment by sgress454** [comment]
@billyshena You can set your log level to `error` or `silent`...http://sailsjs.org/documentation/anatomy/my-app/config/log-js

**Comment by cyrilchapon** [comment]
@sgress454, when I pistes this topic up, I personaly didn't see any further trouble, that was all about the warning

**Comment by wedneyyuri** [comment]
@sgress454 Thanks for your answer. 

I didn't found any errors, but as the common solution is simply ignore the warning doesn't make sense to me pollute the logs .

**Comment by sailsbot** [comment]
Thanks for posting, @boogerlad.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by richdunajewski** [comment]
I still think this should be reopened and fixed. At bare minimum the warning should be reduced to a one-line message so it's not burdening the logging system. The fact that a stack trace is embedded is causing fellow developers to panic and treat these messages as a show-stopping bug/error, when it's really benign in our context.

If we were to extend the session by keeping it alive from the frontend, it would defeat our automatic logout logic that kills the user's session after x minutes of true inactivity (no browsing).

**Comment by sijovijayan** [comment]
+1
`publishDestroy` method not working with in `afterDisconnect` method. How can i fix it ?
It was working fine before with `onDisconnect` method.

I was to do `publishDestroy` only when a socket disconnected.

Please help me.
